packages/middleware-sdk-s3/src/encode-copy-source.ts

```diff
@@ -0,0 +1,34 @@
+import { HandlerArguments, SerializeHandler } from "@aws-sdk/types";
+import { escapeUri, escapeUriPath } from "@aws-sdk/util-uri-escape";
+/**
+ * reference: https://github.com/aws/aws-sdk-ruby/issues/969#issuecomment-151270426
+ */
+export function encodeCopySource(next: SerializeHandler<any, any, any>) {
+  return ({ input, ...others }: HandlerArguments<any>) => {
+    if (input.CopySource) {
+      const copySource: string = input.CopySource;
```

Comment:
Assuming "/" is a separator and occurs just once, this code can be simplified as follows:
```javascript
const [bucket, keyFromSource] = input.CopySource.split("/");
const [key, versionId] = keyFromSource.split("?versionId=");
input.CopySource = `${bucket}/${escapeUriPath(key)}${
  versionId ? `?versionId=${escapeUri(versionId)}` : ""
}`;
```

Reply:
It can have multiple '/'. But we only care the first one. Anything comes after the first `/` is the key and version. 

Your suggestion `keyFromSource.split("?versionId=");` can result in `key = ''`, which is invalid. In this case, the `key` should be `?versionId=XXX`, and it's a valid S3 key. I do this to ensure that when we parse the `versionId`, we should already have a non-empty key.

Reply:
But I agree this part is not easy to read. I'm going to refactor it a bit.

Reply:
Any UTF-8 character is allowed in an object key name ([documentation](https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html#object-key-guidelines))
So by definition, the string `blah?versionId=4` can entirely be key on it's own

Reply:
As described in comment https://github.com/aws/aws-sdk-js-v3/pull/282#issuecomment-506870136, we're not planning to go ahead with this because of the complexity involved.

The key `versionId` is the only one present right now, but new keys might get introduced in future. The URI encoding should be done before CopySource is created as documented at https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#copyObject-property